commit 762a3ec064d0
Author: Emilio Cobos Álvarez <emilio@crisal.io>
Date:   Mon Sep 11 09:51:30 2017 +0200

    Bug 1398581: Ensure a first-letter doesn't incorrectly inherit from a first-line. r=heycam
    
    This approach kinda sucks, because we woefully throw away the style context
    computed in the case it inherited from a first-line, but it's the easiest thing
    I could think of without either making it more inefficient, or threading a
    parent style ignoring first-line through ResolvePseudoElementStyle and related
    functions.
    
    MozReview-Commit-ID: 3uGuU7dHEnE
---
 layout/base/nsCSSFrameConstructor.cpp | 20 +++++++++++++++++---
 1 file changed, 17 insertions(+), 3 deletions(-)

diff --git layout/base/nsCSSFrameConstructor.cpp layout/base/nsCSSFrameConstructor.cpp
index 76d97a32ee4d..b668a837e2f3 100644
--- layout/base/nsCSSFrameConstructor.cpp
+++ layout/base/nsCSSFrameConstructor.cpp
@@ -11861,10 +11861,11 @@ nsCSSFrameConstructor::CreateLetterFrame(nsContainerFrame* aBlockFrame,
 
   // Get style context for the first-letter-frame.  Keep this in sync with
   // nsBlockFrame::UpdatePseudoElementStyles.
-  nsStyleContext* parentStyleContext =
+  nsIFrame* parentFrame =
     nsFrame::CorrectStyleParentFrame(aParentFrame,
-                                     nsCSSPseudoElements::firstLetter)->
-      StyleContext();
+                                     nsCSSPseudoElements::firstLetter);
+
+  nsStyleContext* parentStyleContext = parentFrame->StyleContext();
 
   // Use content from containing block so that we can actually
   // find a matching style rule.
@@ -11873,7 +11874,20 @@ nsCSSFrameConstructor::CreateLetterFrame(nsContainerFrame* aBlockFrame,
   // Create first-letter style rule
   RefPtr<nsStyleContext> sc =
     GetFirstLetterStyle(blockContent, parentStyleContext);
+
   if (sc) {
+    if (sc->IsServo() && parentFrame->IsLineFrame()) {
+      ServoStyleContext* parentStyleIgnoringFirstLine =
+        aBlockFrame->StyleContext()->AsServo();
+      sc =
+        mPresShell->StyleSet()->AsServo()->ReparentStyleContext(
+          sc->AsServo(),
+          parentStyleContext->AsServo(),
+          parentStyleIgnoringFirstLine,
+          parentStyleIgnoringFirstLine,
+          blockContent->AsElement());
+    }
+
     RefPtr<nsStyleContext> textSC = mPresShell->StyleSet()->
       ResolveStyleForText(aTextContent, sc);
 
commit 2a9742d4b68a
Author: Emilio Cobos Álvarez <emilio@crisal.io>
Date:   Mon Sep 11 17:19:05 2017 +0200

    Bug 1398581: Address followup comments. r=me
    
    MozReview-Commit-ID: CymOOl4VLPj
---
 layout/base/nsCSSFrameConstructor.cpp | 15 +--------------
 1 file changed, 1 insertion(+), 14 deletions(-)

diff --git layout/base/nsCSSFrameConstructor.cpp layout/base/nsCSSFrameConstructor.cpp
index b668a837e2f3..bd3d406f664f 100644
--- layout/base/nsCSSFrameConstructor.cpp
+++ layout/base/nsCSSFrameConstructor.cpp
@@ -11884,7 +11884,7 @@ nsCSSFrameConstructor::CreateLetterFrame(nsContainerFrame* aBlockFrame,
           sc->AsServo(),
           parentStyleContext->AsServo(),
           parentStyleIgnoringFirstLine,
-          parentStyleIgnoringFirstLine,
+          parentStyleContext->AsServo(),
           blockContent->AsElement());
     }
 
commit 9a42394a1f71
Author: Emilio Cobos Álvarez <emilio@crisal.io>
Date:   Mon Sep 11 17:19:30 2017 +0200

    Bug 1398581: Crashtest. r=me
    
    MozReview-Commit-ID: 5xy5aeDWU12
---
 layout/style/crashtests/1398581.html    | 17 +++++++++++++++++
 layout/style/crashtests/crashtests.list |  1 +
 2 files changed, 18 insertions(+)

diff --git layout/style/crashtests/1398581.html layout/style/crashtests/1398581.html
new file mode 100644
index 000000000000..e056ab44d55c
--- /dev/null
+++ layout/style/crashtests/1398581.html
@@ -0,0 +1,17 @@
+<script>
+function start() {
+	o3=document.createElement('div');
+	document.body.appendChild(o3);
+	o14=document.createElement('style');
+	document.documentElement.appendChild(o14);
+	o18=document.createElement('style');
+	o14.appendChild(o18);
+	s4=unescape('%u06A10');
+	o3.appendChild(document.createTextNode(s4));
+	o59=document.createTextNode("{}:first-letter{ all: inherit;'x'}\n*{ float: left}:first-line{");
+	o18['before'](o18,-1,o59);
+	document.documentElement.offsetHeight;
+	o3.appendChild(document.createTextNode("x"));
+}
+</script>
+<body onload="start()"></body>
diff --git layout/style/crashtests/crashtests.list layout/style/crashtests/crashtests.list
index a32d48fa3299..719bdaf8328a 100644
--- layout/style/crashtests/crashtests.list
+++ layout/style/crashtests/crashtests.list
@@ -201,3 +201,4 @@ load 1384824-2.html
 load 1387481-1.html
 load 1387499.html
 load 1391577.html
+load 1398581.html
